shared/GIMessage.js

```diff
@@ -74,7 +114,78 @@ export class GIMessage {
   serialize (): string { return this._mapping.value }
 
   hash (): string { return this._mapping.key }
-  // signWithKey (key: Uint8Array) { this._sig = nacl.sign.detached(this.toProtobuf(), key) }
-  // get signature (): Buffer { return this._sig }
-  // set signature (sig: Buffer) { this._sig = sig }
+}
+
+sbp('sbp/selectors/register', {
+  // https://www.wordnik.com/words/chelonia
+  'chelonia/message/process': function (message: GIMessage, state: Object) {
+    sanityCheck(message)
+    const [opT, opV] = message.op()
+    const hash = message.hash()
+    const contractID = message.contractID()
+    const config = sbp('okTurtles.data/get', 'CHELONIA_CONFIG')
+    if (!state._vm) state._vm = {}
+    const opFns = {
+      [GIMessage.OP_CONTRACT] (v) {
+        if (!state._vm.authorizedKeys) state._vm.authorizedKeys = []
+        state._vm.authorizedKeys.push({ key: v.authkey, context: 'owner' })
+      },
+      [GIMessage.OP_ACTION] (v) {
+        if (!config.skipActionProcessing) {
+          const decrypted = message.decrypted(config.decryptFn)
+          opFns[GIMessage.OP_PUB_ACTION](decrypted)
+        }
+      },
+      [GIMessage.OP_PUB_ACTION] (v) {
+        if (!config.skipActionProcessing) {
+          const { data, meta, type } = v
+          if (!config.whitelisted(type)) {
+            throw new Error(`chelonia: action not whitelisted: '${type}'`)
+          }
+          sbp(type, { data, meta, hash, contractID }, state)
+        }
+      },
+      [GIMessage.OP_PROP_SET] (v) {
+        if (!state._vm.props) state._vm.props = {}
+        state._vm.props[v.key] = v.value
+      },
+      [GIMessage.OP_KEY_AUTH] (v) {
+        if (!state._vm.authorizedKeys) state._vm.authorizedKeys = []
+        state._vm.authorizedKeys.push(v)
+      }
+    }
+    let processOp = true
+    if (config.preOp) {
+      processOp = config.preOp(message, state) !== false && processOp
+    }
+    if (config[`preOp_${opT}`]) {
+      processOp = config[`preOp_${opT}`](message, state) !== false && processOp
+    }
+    if (processOp && !config.skipProcessing) {
+      opFns[opT](opV)
+      config.postOp && config.postOp(message, state)
+      config[`postOp_${opT}`] && config[`postOp_${opT}`](message, state)
+    }
+  }
+})
+
+function defaultSignatureFn (data: string) {
+  return {
+    type: 'default',
+    sig: blake32Hash(data)
+  }
+}
+
+function sanityCheck (msg: GIMessage) {
+  const [type] = msg.message().op
+  switch (type) {
+    case GIMessage.OP_CONTRACT:
+      if (!msg.isFirstMessage()) throw new Error('OP_CONTRACT: must be first message')
+      break
+    case GIMessage.OP_ACTION:
+      // nothing forn ow
```

Comment:
forn ow?

Reply:
Oops. Meant "nothing for now", nice catch, will be fixed in the next commit, thanks!